Allow notebook to run as the UID of a local user

alastair opened the pull request:
When running jupyter as the root user, file operations are done as root. This means that any time a file is written on linux, it will be owned as root instead of by the current user.
Make some changes to allow the notebook to run as any local userid.

Also clean up some package installation steps which left behind additional cache files.

alastair committed:
Make all github source dependencies https

alastair committed:
Move `all` target to the top to become the default

alastair committed:
Move Tini installation and fix apt install to delete cache

alastair committed:
Don't cache python installation

alastair committed:
Run jupyter as a custom user, whose id can be changed with an env var

This allows us to run as the same UID as the host user, which
resolves any permissions issues resulting from running as root

alastair committed:
Clean up unclear statements in README

ffont reviewed (approved):
At some point we should look at unifying this image and the one in https://github.com/MTG/asplab-configuration/blob/master/jupyterhub/notebook_dockerfiles/Dockerfile.mtg and make sure the image can run both stand alone or in our JupyterHub deployment in ASPLAB machines. But this is to be discussed in another issue/PR.

alastair committed:
Use COPY instead of ADD (best practises), and chown the password file

alastair committed:
Use gosu instead of sudo to drop to mir user

This prevents forking a new process, as described at
https://github.com/tianon/gosu
